std: Add Motor OS std library port

lasiotus opened the pull request:
Motor OS was added as a no-std Tier-3 target in
[PR 146848](https://github.com/rust-lang/rust/pull/146848) as x86_64-unknown-motor.

This PR adds the std library for Motor OS.

While the PR may seem large, all it does is proxy
std pal calls to [moto-rt](https://crates.io/crates/moto-rt). Where there is some non-trivial
code (e.g. thread::spawn), it is quite similar, often
identical, to what other platforms do.

lasiotus committed:
tidy: allow stdlib to depend on moto-rt

As part of work to add stdlib support for Motor OS.

lasiotus commented:
> Is there any notable std functionality that isn't supported on this platform? Just skimming it, it seems like things are pretty complete (threads, systemtime, process, etc)

I believe all major pieces are supported/implemented, including networking. Some specific things are not (yet) fully implemented, like FS links, or DNS lookup (at the moment, only "localhost" resolves to an IP address). On the other hand, [tokio is ported](https://github.com/moturus/tokio/tree/tokio-motor-1.47.1) via a [mio port](https://github.com/moturus/mio/tree/mio-motor-v1.0.4), which was not completely trivial...

lasiotus commented on library/std/src/sys/fs/mod.rs:
```diff
@@ -27,6 +27,9 @@ cfg_select! {
         mod hermit;
         use hermit as imp;
     }
+    target_os = "motor" => {
+        use crate::sys::pal::fs as imp;
+    }
```
> Note that there is a WIP implementation to migrate the pipe bits over, which this races with. If you want to avoid worrying about that, you can probably drop the pipe API from this PR (just use the unsupported fallback) and just add it back later.

Another rant coming... This one is on a different topic vs the previous one, so I don't feel bound by my promise to rant quietly... :)

+cc @workingjubilee who seems to somewhat share this "too much Unix in Rust" point of view.

&lt;rant&gt;
Pipes seem to me to belong to a Unix world, not Rust std library. Pipes are used to pipe stdio in shell commands, right? Which is Unix. Why are they part of Rust std, while e.g. for an async runtime (which is very much part of the Rust ecosystem) developers have to use a third-party crate? (Not that I want tokio to be in std - tokio seems to be very over-engineered... But that's a different topic).
&lt;/rant&gt;

P.S. I'll see if I can just "drop the pipe API": stdio seems to be tightly coupled with pipes in Rust std, and process/command API also can't live without stdio... Which is probably OK, but still seems very much libc/unix; I hope Rust will come up with something much cleaner/simpler, similar to how async Rust is quite different from what is available in libc... Ranting again? Let me know if you'd like me to tone it down.

lasiotus replied:
Moved everything from sys::pal::motor::[subsystem].rs to sys::[subsystem]::motor.rs, I believe.

bjorn3 commented on library/std/src/panicking.rs:
```diff
@@ -626,7 +626,7 @@ pub fn panicking() -> bool {
 
 /// Entry point of panics from the core crate (`panic_impl` lang item).
-#[cfg(not(any(test, doctest)))]
+#[cfg(all(not(any(test, doctest)), not(target_os = "motor")))]
 #[panic_handler]
```
Looks like moto_rt has it. It writes a panic message (to stderr if it is a terminal, or to the kernel if not), sleeps 100us and then exits with -1 as exit code.

Is there any reason this can't use the std panic handler? Using a custom panic handler means that `std::panic::set_hook` doesn't work.

posk-io commented on library/std/src/os/fd/raw.rs:
```diff
@@ -23,12 +27,15 @@ use crate::sys_common::{AsInner, FromInner, IntoInner};
 
+#[stable(feature = "rust1", since = "1.0.0")]
+#[cfg(target_os = "motor")]
+pub type RawFd = i32;
```
Done.

lasiotus commented on library/std/Cargo.toml:
```diff
@@ -70,6 +70,9 @@ fortanix-sgx-abi = { version = "0.6.1", features = [
     'rustc-dep-of-std',
 ], public = true }
 
+[target.'cfg(target_os = "motor")'.dependencies]
+moto-rt = { version = "0.14", features = ['rustc-dep-of-std'], public = true }
```
Done, thanks.

lasiotus commented on library/std/src/os/motor/mod.rs:
```diff
@@ -0,0 +1,41 @@
+use crate::sealed::Sealed;
+use crate::sys_common::AsInner;
+
+#[unstable(feature = "motor_ext", issue = "none")]
+pub fn map_motor_error(err: moto_rt::ErrorCode) -> crate::io::Error {
```
If a Rust program calls into moto-sys (the Motor OS kernel API) or moto-rt (the Motor OS runtime API), they get ```moto_rt::ErrorCode```. To convert it into io::Error (e.g. for Tokio) they need access to a similar function. So either the function should be exported here as ```pub```, or duplicated elsewhere, from where std can't import (due to io::Error being in std). If io::Error is moved to core (which I've seen being discussed), I'll move this function to moto-rt.

I prefer to avoid code duplication and just export it here. If it is better to duplicate it elsewhere, I can do that also, let me know.

lasiotus replied:
> > I think what you could do here is define the `RawOsError` type as `moto_rt::ErrorCode` (copy how we do this for UEFI) and override `fn decode_error_kind`. That way `io::Error` is actually just a a `moto_rt::ErrorCode`, and [`io::Error::from_raw_os_error`](https://doc.rust-lang.org/std/io/struct.Error.html#method.from_raw_os_error) can be used rather than needing this function.
> 
> Oh, that's a nice trick! I've had a vague desire to use a proper Error enum in public Motor OS API (which is in Rust), and to use a plain ErrorCode/u16 only in internal ABI (which uses C/Linux calling conventions, as Rust doesn't define an ABI).
> 
> So I'm going to spend the next several days refactoring the whole Motor OS codebase around a new enum Error (repr u16, most likely).

Bummer... Spent the weekend to make Motor OS API (moto-rt, moto-sys) return an enum Error instead of a u16, only to find out that Rust library [assumes it is an integer of no more than 32 bits](https://github.com/rust-lang/rust/blob/4fa824bb78318a3cba8c7339d5754b4909922547/library/std/src/io/error/repr_bitpacked.rs#L177). Will just use u16 then - this is a fight for another day...

lasiotus replied:
Tried using u16 as `RawOsError` - downstream crates (getrandom, rand-core) wouldn't compile for Motor OS because they [assume i32](https://github.com/rust-random/rand/blob/4cbd428e88cbe8d7a6e082a5f12c0fdd3aae2759/rand_core/src/os.rs#L78). So Motor OS is using the default `RawOsError` (i32) and overwrites `decode_error_kind`. `map_motor_error` is now crate-internal (and re-uses `decode_error_kind`).

As a T3 target, Motor OS is unstable and so it will be relatively easy to change this later, to e.g. an enum repr(u16), once the ecosystem changes to using RawOsError instead of i32. But RawOsError is not stabilized, so it may take some time...

lasiotus commented on library/std/src/sys/args/motor.rs:
```diff
@@ -0,0 +1,13 @@
+use crate::ffi::OsString;
+
+pub fn args() -> Args {
+    let motor_args = moto_rt::process::args(); // Vec<String>
```
Done.

lasiotus commented on library/std/src/sys/env/motor.rs:
```diff
@@ -0,0 +1,26 @@
+use crate::io;
+
+pub fn env() -> Env {
+    let motor_env = moto_rt::process::env(); // Vec<(String, String)>
```
Done.

lasiotus commented on library/std/src/sys/personality/mod.rs:
```diff
@@ -28,6 +28,12 @@ cfg_select! {
+        fn rust_eh_personality() {
+            core::intrinsics::abort()
+        }
+    }
```
Done.

lasiotus commented on library/std/src/sys/pal/motor/mod.rs:
```diff
@@ -0,0 +1,55 @@
+// NB. used by both libunwind and libpanic_abort
+pub extern "C" fn __rust_abort() {
+    moto_rt::process::exit(-1)
+}
```
Removed, thanks!

lasiotus commented on library/std/src/sys/pal/motor/mod.rs:
```diff
@@ -0,0 +1,55 @@
+    moto_rt::process::exit(-1)
+}
+
+pub(crate) use crate::os::motor::map_motor_error;
```
I replied to the previous comment on this function. I can move it here and re-export in os::motor; or just move it here if you don't buy my argument re: why it is nice to have it exported. Please let me know.

lasiotus commented on library/std/src/sys/pal/motor/common.rs:
Done.

lasiotus commented on library/std/src/sys/pal/motor/os.rs:
```diff
@@ -0,0 +1,92 @@
+}
+
+pub fn current_exe() -> io::Result<PathBuf> {
+    Ok(crate::sys::args::args().next().unwrap().into())
```
In many modern systems, even unices, there is often no global "absolute path", as programs run in their own jails/containers/whatever. Also just by looking at how other platforms deal with it, it is all over the place:

* aix just ["infers"](https://github.com/rust-lang/rust/blob/54a8a1db604e4caff93e26e167ad4a6fde9f0681/library/std/src/sys/pal/unix/os.rs#L261) the location if args[0] is not an absolute path (which is buggy and thus worse, I'd say);
* openbsd does [the same thing](https://github.com/rust-lang/rust/blob/54a8a1db604e4caff93e26e167ad4a6fde9f0681/library/std/src/sys/pal/unix/os.rs#L379) as Motor OS, only additionally calls ```canonicalize()``` if args[0] starts with a dot or contains a slash; if not, args[0] is returned unmodified;
* same for [vxworks](https://github.com/rust-lang/rust/blob/54a8a1db604e4caff93e26e167ad4a6fde9f0681/library/std/src/sys/pal/unix/os.rs#L500).

So I'd say Motor OS is not unique here. I can add a call to ```canonicalize()```, if you insist, but I don't think it will make things better: I'd rather have an unmodified args[0] rather than have it massaged in some cases and not in others.

lasiotus replied:
Added a runtime `fn current_exe` to moto-rt that is now called here. At the moment it does not do anything different, but it will be easy to change later without touching the Rust std library.

lasiotus commented on library/std/Cargo.toml:
```diff
@@ -153,6 +156,7 @@ test = true
 [lints.rust.unexpected_cfgs]
 level = "warn"
 check-cfg = [
+    'cfg(target_os, values("motor"))',
```
Fixed.

Amanieu commented:
We discussed this in the @rust-lang/libs meeting and are happy to add std support for this target because it supports all major functionality of std.

lasiotus commented on library/std/src/os/motor/process.rs:
```diff
@@ -0,0 +1,16 @@
+}
+
+#[unstable(feature = "motor_ext", issue = "none")]
+impl ChildExt for crate::process::Child {
```
Done: https://github.com/rust-lang/rust/issues/147456

lasiotus commented on library/std/src/os/motor/ffi.rs:
Done.

lasiotus commented on library/std/src/sys/io/mod.rs:
Done.

lasiotus commented on library/std/src/sys/pal/motor/os.rs:
```diff
@@ -0,0 +1,91 @@
+    // The CWD is a bad/outdated/unix design, from a single-threaded era:
+    // concurrent changes to CWD lead to races. Applications/processes
+    // should manage their CWD, not the OS.
+    moto_rt::fs::getcwd().map(|s| -> PathBuf { s.into() }).map_err(map_motor_error)
```
Done, thanks!

lasiotus commented on library/std/src/sys/process/motor.rs:
```diff
@@ -0,0 +1,322 @@
+
+        let mut env = Vec::<(String, String)>::new();
+        for (k, v) in self.env.iter() {
+            let val = if let Some(v) = v { v.to_str().unwrap() } else { "" };
+            env.push((k.to_str().unwrap().to_owned(), val.to_owned()));
```
Done.

lasiotus commented on library/std/src/sys/pal/motor/os.rs:
```diff
@@ -0,0 +1,91 @@
+    } else {
+        Err(io::Error::new(io::ErrorKind::InvalidFilename, ""))
+    }
+}
```
A lot of unixy tools, like shells, use "current directory" logic, and the Rust ecosystem is kinda big on those. I'd keep it for compatibility reasons.

lasiotus committed:
Add Motor OS std library port

Motor OS was added as a no-std Tier-3 target in
https://github.com/rust-lang/rust/pull/146848
as x86_64-unknown-motor.

This patch/PR adds the std library for Motor OS.

While the patch may seem large, all it does is proxy
std pal calls to moto-rt. When there is some non-trivial
code (e.g. thread::spawn), it is quite similar, and often
identical, to what other platforms do.

lasiotus commented on library/std/src/sys/net/connection/motor.rs:
```diff
@@ -0,0 +1,529 @@
+                moto_rt::E_OK => None,
+                e => Some(map_motor_error(e)),
+            })
+            .map_err(map_motor_error)
```
Done, thanks!

lasiotus commented:
I believe I have addressed all comments now. Thanks for the detailed review, the PR is definitely better now than what it was two weeks ago!

tgross35 reviewed (changes requested):
Looks pretty good for me! Just a few small requests then I think this should be set.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147000/f6aa851dba4b4509ab7e4f3a407b0b1e72a89e4c..41b36cdea9706c7d300a463f993b4ec03d8fd3bf)*

lasiotus commented on library/std/src/os/motor/ffi.rs:
```diff
@@ -0,0 +1,43 @@
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl OsStrExt for OsStr {
```
Removed the inner tags altogether, as the compiler said the module-level tag is incompatible with inner tags referring to the same issue.

lasiotus commented on library/std/src/sys/process/motor.rs:
```diff
@@ -0,0 +1,324 @@
+impl Command {
+    pub fn new(program: &OsStr) -> Command {
+        let mut env = CommandEnv::default();
+
+        // While it may look like we are iterating over `env` and adding
+        // some of the elements into it again, which is weird, in
+        // reality `env` is empty initially, and env.capture() just
+        // captures the environment without changing `env` var.
+        for (k, v) in env.capture() {
+            if k == moto_rt::process::STDIO_IS_TERMINAL_ENV_KEY {
+                continue;
+            }
+            env.set(&k, &v);
+        }
```
Done. Had to capture() on use below.

thaliaarchi commented on library/std/src/os/motor/ffi.rs:
```diff
@@ -0,0 +1,37 @@
+impl OsStringExt for OsString {
+    #[inline]
+    fn as_str(&self) -> &str {
+        self.to_str().unwrap()
```
I wrote more about my thoughts on this in https://github.com/rust-lang/rust/pull/147797
